crates/wat/src/lib.rs

```diff
@@ -221,6 +222,73 @@ fn _parse_str(wat: &str) -> Result<Vec<u8>> {
     ast.encode().map_err(|e| Error::cvt(e, wat))
 }
 
+/// Result of [`Detect::from_bytes`] to indicate what some input bytes look
+/// like.
+#[derive(Debug, PartialEq, Eq, Clone, Copy)]
+pub enum Detect {
+    /// The input bytes look like the WebAssembly text format.
+    WasmText,
+    /// The input bytes look like the WebAssembly binary format.
+    WasmBinary,
+    /// The input bytes don't look like WebAssembly at all.
+    Unknown,
+}
+
+impl Detect {
+    /// Detect quickly if supplied bytes represent a Wasm module,
+    /// whether binary encoded or in WAT-encoded.
+    ///
+    /// This briefly lexes past whitespace and comments as a `*.wat` file to see if
+    /// we can find a left-paren. If that fails then it's probably `*.wit` instead.
+    ///
+    ///
+    /// Examples
+    /// ```
+    /// use wat::Detect;
+    ///
+    /// assert_eq!(Detect::from_bytes(r#"
+    /// (module
+    ///   (type (;0;) (func))
+    ///   (func (;0;) (type 0)
+    ///     nop
+    ///   )
+    /// )
+    /// "#), Detect::WasmText);
+    /// ```
+    pub fn from_bytes(bytes: impl AsRef<[u8]>) -> Detect {
+        if bytes.as_ref().starts_with(b"\0asm") {
+            return Detect::WasmBinary;
+        }
+        let text = match std::str::from_utf8(bytes.as_ref()) {
+            Ok(s) => s,
+            Err(_) => return Detect::Unknown,
+        };
+
+        let lexer = Lexer::new(text);
+        let mut iter = lexer.iter(0);
+
+        while let Some(next) = iter.next() {
+            match next.map(|t| t.kind) {
+                Ok(TokenKind::Whitespace)
+                | Ok(TokenKind::BlockComment)
+                | Ok(TokenKind::LineComment) => {}
+                Ok(TokenKind::LParen) => return Detect::WasmText,
```

Comment:
Should this check go further to see if the next token is `module` or `component`, or would that be too specific?

Reply:
Ah meant to write more here but got distracted, so wanted to fill this in before I merge. I read this and thought as well, "that's a good idea!", but [according to this](https://webassembly.github.io/spec/core/text/modules.html):

> In a source file, the toplevel surrounding the module body may be omitted.

Meaning that we'd have to support more than just `module` or `component` here but any top-level item, so it's probably best to keep this as-is and see if it causes issue and adapt as necessary from there.